File: algortihms/src/com/algo/dnc/InversionCount.java
package com.algo.dnc;

public class InversionCount {


    public static void main(String[] args) {
        int[] arr = {5,4,3,2,1};
                //{1, 20, 6, 4, 5};
        Inversion inversion = new Inversion(arr);
        System.out.println("Number of inversion = " + inversion.countInversions());


    }


    private static class Inversion {
        private int[] arr;
        private int[] temp;

        Inversion(int[] arr) {
            this.arr = arr;
            this.temp = new int[arr.length];

        }

        private int countInversions() {
            int start = 0;
            int end = arr.length - 1;
            return mergeSort(start, end);

        }

        private int mergeSort(int start, int end) {
            int inversionCount = 0;
            if (start < end) {
                int mid = (start + end) / 2;
                inversionCount = mergeSort(start, mid);
                inversionCount += mergeSort(mid + 1, end);
                inversionCount += merge(start, mid, end);

            }
            return inversionCount;
        }

        private int merge(int start, int mid, int end) {

            int i = start;
            int j = mid + 1;
            int k = start;
            int inversionCount = 0;
            while ((i <= mid) && (j <= end)) {
                if (arr[i] <= arr[j]) {
                    temp[k] = arr[i];
                    ++i;
                } else {
                    temp[k] = arr[j];
                    ++j;
                    inversionCount = inversionCount + (mid - i + 1);
                }
                ++k;
            }

            while (i <= mid) {
                temp[k++] = arr[i];
                ++i;
            }
            while (j <= end) {
                temp[k++] = arr[j];
                ++j;
            }
            for (i = start; i <= end; i++) {
                arr[i] = temp[i];
            }
            return inversionCount;
        }
    }
}

File: algorithms-python/dnc/inversion_count.py
from typing import List


class InversionCount:
    def count_inversion(self, arr: List[int]) ->int:
        return self._count_inversion(arr, 0, len(arr) - 1)
    def _count_inversion(self, arr: List[int], start: int, end: int)->int:
        if start >= end:
            return 0
        count = 0
        mid = (start + end) // 2
        count += self._count_inversion(arr, start, mid)
        count += self._count_inversion(arr, mid + 1, end)
        count += self._merge(arr, start, mid, end)
        return count
    
    def _merge(self, arr: List[int], start: int, mid: int, end: int) -> int:
        temp: List[int] = []
        i = start
        j = mid + 1        
        count = 0
        while (i <= mid) and (j <= end):
            if arr[i] <= arr[j]:
                temp.append(arr[i])
                i += 1
            else:
                temp.append(arr[j])
                count = count + (mid - i + 1)    
                j += 1
        while i <= mid:
            temp.append(arr[i])
            i += 1
        while j <= end:
            temp.append(arr[j])
            j += 1
        for i in range(len(temp)):
            arr[i + start] = temp[i]
        return count


def main() -> None:
    nums = [5,4,3,2,1]
    inversion_count = InversionCount()
    print(f"Inverson count = {inversion_count.count_inversion(nums)}")


if __name__ == "__main__":
    main()

               
